lib/NetworkManager.js

```diff
@@ -61,6 +61,14 @@ class NetworkManager extends EventEmitter {
     await this._updateProtocolRequestInterception();
   }
 
+  /**
+   * @param {?{username: string, password: string}} credentials
+   */
+  async authenticateProxy(credentials) {
```

Comment:
we might need to come up with a different shape for the API. WIth this approach, we'll have:
- `page.authenticate` to setup credentials for both server and proxy auth challenges
- `page.authenticateProxy` to setup credentials for proxy only

It's unclear from the api how to setup auth for server-only, or how to setup different creds for server and proxy.

maybe we should add a "source" parameter to `page.authenticate` object to explicitly set auth source?

Reply:
@aslushnikov I would argue in favor of the existing API, as authenticating a web page and authenticating a proxy server are conceptually different tasks.

But I don't mind adding a `source` field. It can even default to web page authentication in case it's not specified.